zerver/tests/test_presence.py

```diff
@@ -87,6 +87,44 @@ def back_date(num_weeks: int) -> None:
         self.assert_length(presence_dct, 0)
         self.assertEqual(last_update_id, -1)
 
+    def test_user_presence_row_creation_simulated_race(self) -> None:
+        """
+        There is a theoretical race condition, where while a UserPresence
+        row is being created for a user, a concurrent process creates it first,
+        right before we execute our INSERT. This conflict is handled with
+        ON CONFLICT DO NOTHING in the SQL query and an early return
+        if that happens.
+
+        Such a race is difficult to replicate in a test, but the least we can do
+        is simulate the conditions that do_update_user_presence uses to detect
+        that the ON CONFLICT DO NOTHING case happened.
+        """
+        UserPresence.objects.all().delete()
```

Comment:
Might as well filter to the target user for this; it's a bit cheaper and a good habit to not perturb the database any more than is necessary.